caddy/app.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package caddy provides an app and handler for Caddy Server (https://caddyserver.com/)
+// allowing it to turn any handler into one supporting the Vulcain protocol.
+
+package caddy
+
+import (
+	"errors"
+	"log/slog"
+
+	outline_prometheus "github.com/Jigsaw-Code/outline-ss-server/prometheus"
+	outline "github.com/Jigsaw-Code/outline-ss-server/service"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func init() {
+	caddy.RegisterModule(OutlineApp{})
+}
+
+const moduleName = "outline"
+
+type ShadowsocksConfig struct {
+	ReplayHistory int `json:"replay_history,omitempty"`
+}
+
+type OutlineApp struct {
+	ShadowsocksConfig *ShadowsocksConfig `json:"shadowsocks,omitempty"`
+
+	ReplayCache outline.ReplayCache
+	logger      *slog.Logger
+	Metrics     outline.ServiceMetrics
+	buildInfo   *prometheus.GaugeVec
+}
+
+func (OutlineApp) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  moduleName,
+		New: func() caddy.Module { return new(OutlineApp) },
+	}
+}
+
+// Provision sets up Outline.
+func (app *OutlineApp) Provision(ctx caddy.Context) error {
+	app.logger = ctx.Slogger()
+
+	app.logger.Info("provisioning app instance")
+
+	if app.ShadowsocksConfig != nil {
+		app.ReplayCache = outline.NewReplayCache(app.ShadowsocksConfig.ReplayHistory)
```

Comment:
Is this created on every config change? We should preserve the replay history across config updates.

Reply:
It is recreated, yes. When Caddy reloads, it terminates the existing instance of the app and starts a new one with the updated config.

We have to manage the persistence of replay history some other way. Caddy provides and suggests using its [`UsagePool`](https://pkg.go.dev/github.com/caddyserver/caddy/v2#UsagePool) to manage state across loads, but it requires using a global.

I was going to look into this next as it may require a small change to `ReplayCache` to allow resizing the cache capacity in case the config changes the value of `replay_history`.